### PR TITLE
Add sync batch norm grad

### PR DESCRIPTION
**Pull request by 404988613**
### PR types
Others

### PR changes
Others

### Describe
61
- #51292

### CODE REVIEW

**Comment by luotao1** [comment]
请解决下冲突

**Comment by 404988613** [comment]
请问一下，这个里面的输出没有显示数据类型，然后只在黑名单中删除对应的算子还是报错，这个该怎么解决呀

**Comment by From00** [comment]
> 请问一下，这个里面的输出没有显示数据类型，然后只在黑名单中删除对应的算子还是报错，这个该怎么解决呀

SyncBatchNormGradKernel调用了SyncBatchNormGradFunctor，你需要从SyncBatchNormGradFunctor中判断出输出Tensor的数据类型。

**Comment by 404988613** [comment]
您好，我现在从SyncBatchNormGradFunctor中判断出输出Tensor1和2的数据类型为BatchNormParamType，然后注册为UNDEFINED了，需要在InferMeta中对输出DataType进行推导，但是我好像在InferMeta中没有找到这个对应的void函数SyncBatchNormGradInferMeta呀

**Comment by From00** [comment]
> 您好，我现在从SyncBatchNormGradFunctor中判断出输出Tensor1和2的数据类型为BatchNormParamType，然后注册为UNDEFINED了，需要在InferMeta中对输出DataType进行推导，但是我好像在InferMeta中没有找到这个对应的void函数SyncBatchNormGradInferMeta呀

这个算子的InferMeta还没迁移到phi下。你也可以在注册的花括号里直接写推导逻辑，参考batch_norm_grad：https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/phi/kernels/gpu/batch_norm_grad_kernel.cu#L1324

**Comment by 404988613** [comment]
您好，可以帮忙看一下问题吗？我参考batch_norm_grad进行了修改，不知道有具体哪些数据类型了

File: paddle/phi/kernels/gpu/sync_batch_norm_grad_kernel.cu [hunk 1]
```diff
@@ -68,5 +68,11 @@ PD_REGISTER_KERNEL(sync_batch_norm_grad,
                    phi::SyncBatchNormGradKernel,
                    float,
                    double,
-                   phi::dtype::float16) {}
+                   phi::dtype::float16) {
+  if (kernel_key.dtype() == phi::DataType::FLOAT16) {
+      kernel->OutputAt(0).SetDataType(phi::DataType::FLOAT32);  // x_grad
```
Comment (From00) [inline paddle/phi/kernels/gpu/sync_batch_norm_grad_kernel.cu]: x_grad输出类型没变，不需要注册。

File: paddle/phi/kernels/sparse/gpu/sync_batch_norm_grad_kernel.cu [hunk 2]
```diff
@@ -72,12 +72,41 @@ PD_REGISTER_KERNEL(sync_batch_norm_coo_grad,
                    phi::sparse::SyncBatchNormCooGradKernel,
                    float,
                    phi::dtype::float16) {}
+
+PD_REGISTER_KERNEL(sync_batch_norm_grad,
```
Comment (From00) [inline paddle/phi/kernels/sparse/gpu/sync_batch_norm_grad_kernel.cu]: 不需要重复注册kernel

Comment (404988613) [reply paddle/phi/kernels/sparse/gpu/sync_batch_norm_grad_kernel.cu]: 删除掉重复注册的kernel了，好像还是有问题，是if中数据类型判断错误了吗

Comment (From00) [reply paddle/phi/kernels/sparse/gpu/sync_batch_norm_grad_kernel.cu]: 你的PR和develop分支存在代码冲突。你需要pull develop分支，解决冲突之后再重新提交。

**Comment by luotao1** [comment]
close due to the following PR is merged:
- #52941
